File: Bike.go

package main

/* Imports
 * 4 utility libraries for formatting, handling bytes, reading and writing JSON, and string manipulation
 * 2 specific Hyperledger Fabric specific libraries for Smart Contracts
 */
import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/hyperledger/fabric/core/chaincode/shim"
	sc "github.com/hyperledger/fabric/protos/peer"
)

// Define the Smart Contract structure
type SmartContract struct {
}

// Define the bike structure, with 6 properties.  Structure tags are used by encoding/json library
type Bike struct {
	Make   string `json:"make"`
	Model  string `json:"model"`
	Colour string `json:"colour"`
	Vin    string `json:"vin"`
	EngineCC string `json:"engcc"`
	Owner  string `json:"owner"`
}

type SmartContract struct {
}

// Define the person structure, with 4 properties.  Structure tags are used by encoding/json library
type Person struct {
	Name   string `json:"name"`
	Address string `json:"address"`
	Email   string `json:"email"`
	Owner  string `json:"owner"`
}




/*
 * The Init method is called when the Smart Contract "fabBike" is instantiated by the blockchain network
 * Best practice is to have any Ledger initialization in separate function -- see initLedger()
 */
func (s *SmartContract) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
	return shim.Success(nil)
}

/*
 * The Invoke method is called as a result of an application request to run the Smart Contract "fabBike"
 * The calling application program has also specified the particular smart contract function to be called, with arguments
 */
func (s *SmartContract) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {

	// Retrieve the requested Smart Contract function and arguments
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "queryBike" {
		return s.queryBike(APIstub, args)
	} else if function == "initLedger" {
		return s.initLedger(APIstub)
	} else if function == "createBike" {
		return s.createBike(APIstub, args)
	} else if function == "queryAllBikes" {
		return s.queryAllBikes(APIstub)
	} else if function == "changeBikeOwner" {
		return s.changeBikeOwner(APIstub, args)
	}

	return shim.Error("Invalid Smart Contract function name.")
}

func (s *SmartContract) queryBike(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 1 {
		return shim.Error("Incorrect number of arguments. Expecting 1")
	}

	BikeAsBytes, _ := APIstub.GetState(args[0])
	return shim.Success(BikeAsBytes)
}

func (s *SmartContract) initLedger(APIstub shim.ChaincodeStubInterface) sc.Response {
	Bikes := []Bike{
		Bike{Make: "Honda", Model: "Activa", Colour: "gold",Vin: "567SDA12",EngineCC: "100CC",Owner: "Sandhya"},
		Bike{Make: "Vespa", Model: "Activa", Colour: "red",Vin: "345VLS67",EngineCC: "150CC" ,Owner: "Lakshmi"},
		Bike{Make: "Bajaj", Model: "AXv", Colour: "blue",Vin: "135AB79",EngineCC: "150CC" ,Owner: "Satwik"},
		Bike{Make:  "HeroHonda", Model: "Infinit", Colour: "grey",Vin: "145HG45",EngineCC: "100CC" , Owner: "Prabhu"},
		Bike{Make:  "Yamaha", Model: "", Colour: "metallic",Vin: "179NF68",EngineCC: "150CC" , Owner: "Syed"},
		Bike{Make:  "TVS", Model: "", Colour: "green",Vin: "236KJ12",EngineCC: "150CC" , Owner: "Anand"},
		Bike{Make: "Honda" , Model: "Kinetic", Colour: "silver",Vin: "934DS21",EngineCC: "150CC" , Owner: "Seshu"},
		Bike{Make:  "", Model: "", Colour: "yellow",Vin: "689LK12",EngineCC: "100CC" , Owner: "Pavan"},
		Bike{Make:  "Honda", Model: "Kinetic", Colour: "black",Vin: "123MHK45",EngineCC: "100CC" , Owner: "Vani"},
		Bike{Make: "", Model: "Pleasure", Colour: "white",Vin: "4566VMS78",EngineCC: "50CC" , Owner: "Madhuri"},
	    Bike{Make:  "", Model: "", Colour: "pink",Vin: "",EngineCC: "100CC" , Owner: "Yogita"},	
	}

	i := 0
	for i < len(Bikes) {
		fmt.Println("i is ", i)
		BikeAsBytes, _ := json.Marshal(Bikes[i])
		APIstub.PutState("Bike"+strconv.Itoa(i), BikeAsBytes)
		fmt.Println("Added", Bikes[i])
		i = i + 1
	}

	return shim.Success(nil)
}

func (s *SmartContract) createBike(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 6 {
		return shim.Error("Incorrect number of arguments. Expecting 6")
	}

	var Bike = Bike{Make: args[1], Model: args[2], Colour: args[3], Vin: args[4],EngineCC: args[5],Owner: args[6]}

	BikeAsBytes, _ := json.Marshal(Bike)
	APIstub.PutState(args[0], BikeAsBytes)

	return shim.Success(nil)
}

func (s *SmartContract) queryAllBikes(APIstub shim.ChaincodeStubInterface) sc.Response {

	startKey := "Bike0"
	endKey := "Bike999"

	resultsIterator, err := APIstub.GetStateByRange(startKey, endKey)
	if err != nil {
		return shim.Error(err.Error())
	}
	defer resultsIterator.Close()

	// buffer is a JSON array containing QueryResults
	var buffer bytes.Buffer
	buffer.WriteString("[")

	bArrayMemberAlreadyWritten := false
	for resultsIterator.HasNext() {
		queryResponse, err := resultsIterator.Next()
		if err != nil {
			return shim.Error(err.Error())
		}
		// Add a comma before array members, suppress it for the first array member
		if bArrayMemberAlreadyWritten == true {
			buffer.WriteString(",")
		}
		buffer.WriteString("{\"Key\":")
		buffer.WriteString("\"")
		buffer.WriteString(queryResponse.Key)
		buffer.WriteString("\"")

		buffer.WriteString(", \"Record\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(string(queryResponse.Value))
		buffer.WriteString("}")
		bArrayMemberAlreadyWritten = true
	}
	buffer.WriteString("]")

	fmt.Printf("- queryAllBikes:\n%s\n", buffer.String())

	return shim.Success(buffer.Bytes())
}

func (t *SimpleChaincode) getBikeDetails(stub shim.ChaincodeStubInterface, args []string) sc.Response{
	
	return shim.Success(nil)

}






func (s *SmartContract) changeBikeOwner(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {

	if len(args) != 2 {
		return shim.Error("Incorrect number of arguments. Expecting 2")
	}

	BikeAsBytes, _ := APIstub.GetState(args[0])
	Bike := Bike{}

	json.Unmarshal(BikeAsBytes, &Bike)
	Bike.Owner = args[0]

	BikeAsBytes, _ = json.Marshal(Bike)
	APIstub.PutState(args[0], BikeAsBytes)

	return shim.Success(nil)
}









// The main function is only relevant in unit test mode. Only included here for completeness.
func main() {

	// Create a new Smart Contract
	err := shim.Start(new(SmartContract))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}

File: Bike.js
/*
# Copyright IBM Corp. All Rights Reserved.
#
# SPDX-License-Identifier: Apache-2.0
*/

'use strict';
const shim = require('fabric-shim');
const util = require('util');

let Chaincode = class {

  // The Init method is called when the Smart Contract 'fabBike' is instantiated by the blockchain network
  // Best practice is to have any Ledger initialization in separate function -- see initLedger()
  async Init(stub) {
    console.info('=========== Instantiated fabBike chaincode ===========');
    return shim.success();
  }

  // The Invoke method is called as a result of an application request to run the Smart Contract
  // 'fabBike'. The calling application program has also specified the particular smart contract
  // function to be called, with arguments
  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);

    let method = this[ret.fcn];
    if (!method) {
      console.error('no function of name:' + ret.fcn + ' found');
      throw new Error('Received unknown function ' + ret.fcn + ' invocation');
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  async queryBike(stub, args) {
    if (args.length != 1) {
      throw new Error('Incorrect number of arguments. Expecting BikeNumber ex: Bike01');
    }
    let BikeNumber = args[0];

    let BikeAsBytes = await stub.getState(BikeNumber); //get the Bike from chaincode state
    if (!BikeAsBytes || BikeAsBytes.toString().length <= 0) {
      throw new Error(BikeNumber + ' does not exist: ');
    }
    console.log(BikeAsBytes.toString());
    return BikeAsBytes;
  }

  async initLedger(stub, args) {
    console.info('============= START : Initialize Ledger ===========');
    let Bikes = [];
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
     make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });
    Bikes.push({
      make: 'Honda',
      model: 'Activa',
      color: 'gold',
      vin: '567SDA12',
      EngineCC:'100cc',
      owner: 'Sandhya'
    });

    for (let i = 0; i < Bikes.length; i++) {
      Bikes[i].docType = 'Bike';
      await stub.putState('Bike' + i, Buffer.from(JSON.stringify(Bikes[i])));
      console.info('Added <--> ', Bikes[i]);
    }
    console.info('============= END : Initialize Ledger ===========');
  }

  async createBike(stub, args) {
    console.info('============= START : Create Bike ===========');
    if (args.length != 5) {
      throw new Error('Incorrect number of arguments. Expecting 5');
    }

    var Bike = {
      docType: 'Bike',
      make: args[1],
      model: args[2],
      color: args[3],
      vin: args[4],
      EngineCC: args[5],
      owner: args[6]
    };

    await stub.putState(args[0], Buffer.from(JSON.stringify(Bike)));
    console.info('============= END : Create Bike ===========');
  }

  async queryAllBikes(stub, args) {

    let startKey = 'Bike0';
    let endKey = 'Bike999';

    let iterator = await stub.getStateByRange(startKey, endKey);

    let allResults = [];
    while (true) {
      let res = await iterator.next();

      if (res.value && res.value.value.toString()) {
        let jsonRes = {};
        console.log(res.value.value.toString('utf8'));

        jsonRes.Key = res.value.key;
        try {
          jsonRes.Record = JSON.parse(res.value.value.toString('utf8'));
        } catch (err) {
          console.log(err);
          jsonRes.Record = res.value.value.toString('utf8');
        }
        allResults.push(jsonRes);
      }
      if (res.done) {
        console.log('end of data');
        await iterator.close();
        console.info(allResults);
        return Buffer.from(JSON.stringify(allResults));
      }
    }
  }

  async changeBikeOwner(stub, args) {
    console.info('============= START : changeBikeOwner ===========');
    if (args.length != 2) {
      throw new Error('Incorrect number of arguments. Expecting 2');
    }

    let BikeAsBytes = await stub.getState(args[0]);
    let Bike = JSON.parse(BikeAsBytes);
    Bike.owner = args[1];

    await stub.putState(args[0], Buffer.from(JSON.stringify(Bike)));
    console.info('============= END : changeBikeOwner ===========');
  }
};

shim.start(new Chaincode());
